atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/locks/Lockable.java

```diff
@@ -0,0 +1,133 @@
+// at all, but with so many threads and relatively high chance of some collisions, it seems like wasted RPCs otherwise)
+public final class Lockable<T> {
+    private static final SafeLogger log = SafeLoggerFactory.get(Lockable.class);
+    private final LockDescriptor lockDescriptor;
+    private final T inner;
+    private final TimelockService timeLock;
+    private final Refreshable<Duration> lockTimeout;
+
+    // Consider whether this should be a lock. It's simply an optimisation to avoid a check for timelock
+    // each time you want to test if it's locked by something locally, but we must make sure there are no false
+    // negatives. It's allowed that something local is working on this item and Timelock has released the lock due to
+    // expiry and this is still True (since that still lets us minimise conflicts by not working on it locally),
+    // but it should not be true if something local has stopped working on this item and closed the inner class
+    private final AtomicBoolean isLocked = new AtomicBoolean(false);
+
+    private Lockable(
+            T inner, LockDescriptor lockDescriptor, TimelockService timeLock, Refreshable<Duration> lockTimeout) {
+        this.timeLock = timeLock;
+        this.inner = inner;
+        this.lockTimeout = lockTimeout;
+        this.lockDescriptor = lockDescriptor;
+    }
+
+    public static <T> Lockable<T> create(
+            T inner, LockDescriptor lockDescriptor, TimelockService timeLock, Refreshable<Duration> lockTimeout) {
+        return new Lockable<>(inner, lockDescriptor, timeLock, lockTimeout);
+    }
+
+    // For code owners:
+    // Do not use this for multiple locks!! If you're not careful, you can make it really easy to deadlock
+    // If you end up using this class and have a need for batch locks, implement a separate method that uses
+    // Timelocks batch locking methods and ensure there's a consistent ordering when taking out local locks
+    // to avoid circular waiting.
+    //    @MustBeClosed TODO: How do I get this working?? I'll track looking into this for another time.
+    @CheckReturnValue
+    public Optional<Inner> tryLock() {
+        if (isLocked.compareAndSet(false, true)) {
+            // We do not want the timeout to be too low to avoid a race condition where we give up too soon
+            LockRequest request = LockRequest.of(
+                    ImmutableSet.of(lockDescriptor), lockTimeout.get().toMillis());
+
+            try {
+                return timeLock.lock(request).getTokenOrEmpty().map(lockToken -> new Inner(lockToken));
+            } catch (Exception e) {
+                log.error("Failed to acquire lock", UnsafeArg.of("lockDescriptor", lockDescriptor), e);
+                isLocked.set(false);
+                return Optional.empty();
+            }
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    // Expected to be used in a try closeable.
+    public final class Inner implements AutoCloseable {
```

Comment:
Do you want an annotation for this? (`@MustBeClosed` or similar?)

Also, we probably shouldn't be using `Inner` / `getInner` / `getOuter` for this class / these methods - seems to be a bit of an implementation detail.

Reply:
See the comment on the tryLock method - yes, we want MustBeClosed, but putting it here doesn't do what we want it to do. I've still gotta figure that one out.

For naming - yes. I left a comment to that effect on Saturday (saying I need to change the naming), but it appears I have rebased it away.